spotpass/scrape-3ds.js

```diff
@@ -0,0 +1,90 @@
+const https = require('node:https');
+const axios = require('axios');
+const fs = require('fs-extra');
+const { COUNTRIES, LANGUAGES } = require('./constants');
+const apps = require('./ctr-boss-apps.json');
+
+const NPFL_URL_BASE = 'https://npfl.c.app.nintendowifi.net/p01/filelist';
+const NPDL_URL_BASE = 'https://npdl.cdn.nintendowifi.net/p01/nsa';
+
+const httpsAgent = new https.Agent({
+	rejectUnauthorized: false,
+	cert: fs.readFileSync('./certs/wiiu-common.crt'), // * Hey, it works lol
+	key: fs.readFileSync('./certs/wiiu-common.key'),
+});
+
+async function scrape3DS(downloadBase) {
+	await Promise.all(COUNTRIES.map(async (country) => {
+		await scrapeCountry(country, downloadBase);
+	}));
+}
+
+async function scrapeCountry(country, downloadBase) {
+	await Promise.all(LANGUAGES.map(async (language) => {
+		await scrapeLanguage(country, language, downloadBase);
+	}));
+}
+
+async function scrapeLanguage(country, language, downloadBase) {
+	for (const app of apps) {
+		for (const task of app.tasks) {
+			const response = await axios.get(`${NPFL_URL_BASE}/${app.app_id}/${task}?c=${country}&l=${language}`, {
+				validateStatus: () => {
+					return true;
+				},
+				httpsAgent
+			});
+
+			if (!response.headers['content-type'] || !response.headers['content-type'].startsWith('text/plain')) {
+				continue;
+			}
+
+			fs.ensureDirSync(`${downloadBase}/${country}/${language}/${app.app_id}/${task}`);
+			fs.writeFileSync(`${downloadBase}/${country}/${language}/${app.app_id}/${task}/filelist.txt`, response.data);
+
+			const lines = response.data.split('\r\n').filter(line => line);
+			const files = lines.splice(2)
+
+			// * There's like 4 ways the 3DS can format these download URLs, just pray this works I guess.
+			// * Not sure any better way to do this.
+			for (const file of files) {
+				const parts = file.split('\t');
+				const fileName = parts[0];
+
+				// * There are 4 possible formats for NPDL URLs.
+				// * This tries all of them, one after the other, from least
+				// * specific to most specific. This should result in the most
+				// * specific version of each file being downloaded, overwriting
+				// * less specific ones. Not all files work with all formats, so
+				// * we just have to try them all and pray.
+				// * This is pretty slow, but it at least should get all the data.
+				const downloadPath = `${downloadBase}/${country}/${language}/${app.app_id}/${task}/${fileName}.boss`;
+
+				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${fileName}`, downloadPath);
+				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${language}/${fileName}`, downloadPath);
+				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${language}_${country}/${fileName}`, downloadPath);
+				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${country}/${language}/${fileName}`, downloadPath);
```

Comment:
maybe we could return early if a less specific version worked with no issues?

Reply:
Ideally an early return would be done the other way around. The goal with trying them all is the get the most specific one as possible. My worry is that there may be cases where a title has a less specific "general" version of the file, which uses some default region, and also region-specific versions of the file

So say a game has an ES-es version of the file, but it also accepts a download of just the file name which defaults to the US-en region. If we only use the less-specific version then we miss the ES-es version of the file

Reply:
I have no proof to back this up, I just know Nintendo really sucks and I'm trying to account for all possible issues

Reply:
in theory, if a less specific version of boss content exists, then the contents should be exactly the same for the more specific versions. this way, if we return early we won't miss anything and use less storage

Reply:
Is this something we want to risk or can we verify it? In the hypothetical example I provided the contents would differ since the default file would be for one region and not the other (it assumes that a task has both versions available and allows for both specific and generic downloads)

I agree with an early return, but I do still think going the other way around may be better? As you said in your docs if the country/language is not enforced (meaning there's only the default file), then they aren't validated at all. So a request using the specific country and language should return early even if the content wouldn't differ either way, while still potentially catching my hypothetical issue

Reply:
If the server has differing content for different regions, then the country and/or language *are* enforced. This means that using the general path will return an error (404 in the case of NPDL), so checking the general path first should be right because it won't work if the country or language are enforced.

BTW aside from all of this, there are some cases where the only argument is the *country* and not the *language*, so this means there are 5 possible options:

```suggestion
				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${fileName}`, downloadPath);
				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${language}/${fileName}`, downloadPath);
				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${country}/${fileName}`, downloadPath);
				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${language}_${country}/${fileName}`, downloadPath);
				await downloadContentFile(`${NPDL_URL_BASE}/${app.app_id}/${task}/${country}/${language}/${fileName}`, downloadPath);
```

Reply:
> If the server has differing content for different regions, then the country and/or language _are_ enforced

The hypothetical I was proposing was that this may not always be the case for some Nintendo:tm: fucky reason. Though thinking about it now, using the less generic ones first may still be slower?

The more specific ones should work in more cases, since they will work in cases both where:

- Country and language *are* enforced
- Country and language *aren't* enforced (since they aren't validated)

That means that they should succeed more times than not, no? If we start with the generic requests first, we may have to make *more* requests in cases where these things are enforced, where as if we start from the more specific ones we should hit a success faster as they aren't validated anyway and just succeed



> BTW aside from all of this, there are some cases where the only argument is the _country_ and not the _language_, so this means there are 5 possible options

Fuck

Reply:
Well yeah you are right, but I was talking about saving space in disk after all tasks are processed to avoid duplicates. We can leave it as it is aswell

Reply:
You can disregard this, I didn't see that the download path was a constant using the country and language

Reply:
:+1: